### PR TITLE
Inline types to compiled module

### PR DESCRIPTION
**Pull request by fishcakez**
So far simple extensions can be inlined to a single module. Encode avoids external calls (where possible) and is more efficient for fixed length types. Decode avoids generating reference binaries and external calls (where possible), does fewer matches and uses the same binary match context as the row for most types.

Will need to review API and benchmark before continuing the work.

### CODE REVIEW

File: lib/postgrex.ex [hunk 1]
```diff
@@ -66,6 +66,13 @@ defmodule Postgrex do
     if not `DBConnection.Connection` (default: `DBConnection.Connection`);
     * `:null` - The atom to use as a stand in for postgres' `NULL` in encoding
     and decoding (default: `nil`);
+    * `:date` - Either `:postgrex` to use Postgrex date structs or
```
Comment (josevalim) [inline lib/postgrex.ex]: @ericmj are we sure we don't want to require Elixir v1.3 by default and default to `:elixir` and keep the `:postgrex` ones only for backwards compatibility?

Comment (ericmj) [inline lib/postgrex.ex]: You are right. It's a good idea to do the backwards incompatible change before 1.0 but still keep a compatible alternative as an option.

Comment (josevalim) [inline lib/postgrex.ex]: Ok, I will open up a separate issue.

File: lib/postgrex/extension.ex [hunk 2]
```diff
@@ -67,14 +90,31 @@ defmodule Postgrex.Extension do
...
+      end
+
   """
-  @callback encode(TypeInfo.t, term, Types.types, opts) :: iodata
+  @callback encode(opts) :: Macro.expr
 
   @doc """
-  Should decode a binary in the specified Postgres protocol format to an Elixir
-  value.
+  Returns a quoted list of clauses that decode a binary to an Elixir value. The
```
Comment (josevalim) [inline lib/postgrex/extension.ex]: The first paragraph must always be a short summary. Just start a paragraph after the first sentence. :)

**Comment by fishcakez** [comment]
If we could compile the module faster I think we could get most tests to pass but things are really taking shape now. It takes about 8 seconds to bootstrap on my machine. Have not looked into what takes the time to be honest.

Some todo's:
* [ ] Add tests for decoding anonymous records
* [ ] Add tests for new code/type servers
* [ ] Support not rebootstrapping
* [ ] Improve compile time type modules support
* [ ] Do some benchmarks using real queries
* [ ] Investigate decoding rows when they are received[1]

[1] We are a lot faster at decoding (even before this PR) than when we benefited from decoding outside in the query callback. Using some trickier we would be able to decode a full buffer of rows with a single binary match context.

**Comment by fishcakez** [comment]
Now we have a compile time types modules that provide an optimised resolver for bootstrapping and we can make use of already bootstrapped types for super extensions. The implementation needs some refactoring and docs to make it usable for custom modules at compile time. Most tests outside of the type server tests are failing because we only use a default types module but we are back to normal testing speed.

@josevalim please could you step in again to work on the public API for `Postgrex.TypeModule`?

**Comment by fishcakez** [comment]
Testing the full connect+handshake+bootstrap, a database taking 4.4 seconds on master takes 2.4 seconds on this branch, and one taking 8.9 seconds on master takes 3.4 seconds on this branch. I guess that shows the O(N*M) -> O(N*log(M)) quite well.

**Comment by fishcakez** [comment]
For a query of 300k rows of length 6 with small int/text/timestamp values this branch, as of cc66059, reduces the query time by 43% compared to master.

**Comment by fishcakez** [comment]
Ah I forgot to revert 52c0309. I will look at that tomorrow.

**Comment by josevalim** [comment]
I will ping you on IRC and discuss how to define a public API for this
today. Amazing work!

**Comment by fishcakez** [comment]
I reverted 52c0309 manually. It buys a couple of percent compared to cc66059 in the benchmark above.

**Comment by fishcakez** [comment]
If we required Elixir ~> 1.3 tests would green up. I resisted fixing the issues as we might be requiring this for 1.0 anyway.

**Comment by fishcakez** [comment]
Oh we get some compile warnings too, not sure why those would occur.

**Comment by josevalim** [comment]
@fishcakez it is easy to fix the generated error by using module attributes. Maybe we should do it?

**Comment by josevalim** [comment]
I believe this is ready to go. It is approved from my side!

**Comment by fishcakez** [comment]
Yeah I think we are ready to go. Please see https://github.com/elixir-ecto/postgrex/pull/263/commits/571fb5849037a3cc891a0a50992c6d948f32da94 for removed functionality.

**Comment by dantswain** [comment]
Hey @fishcakez I'm working on converting #265 to use this new approach.  In the intermediate state, when things are broken and I expect them to be broken, I'm getting some not-so-helpful compiler errors.  I don't know if there's anything that can be _done_ about it, but I thought you might be interested to see it.  

I have a bug in one of my decode pattern matches in the polygon extension, and the error message the compiler gives me is:

```
== Compilation error on file lib/postgrex/default_types.ex ==
** (CompileError) lib/postgrex/type_module.ex:716: a binary field without size is only allowed at the end of a binary pattern
    (stdlib) lists.erl:1338: :lists.foreach/2
    (elixir) lib/kernel/parallel_compiler.ex:117: anonymous fn/4 in Kernel.ParallelCompiler.spawn_compilers/1
```

This is the `decode/1` clause:

```
  def decode(_) do
    quote location: :keep do
      <<length::int32, rest::bits>> ->
        # TODO
        %Postgrex.Polygon{vertices: []}
    end
  end
```

My guess on why I'm getting an error is now you're taking the match clause and inserting it into a bigger binary match?  I should be able to specify the actual number of bytes, which will be better anyways.  The point was mainly that the error message doesn't really tell me where the problem is in the code.

**Comment by dantswain** [comment]
Another interesting (kind of confusing) error message.  This one is at run-time during tests.  My binary match is still mucked up I'm sure, but I don't know where the "Deprecated" came from.  The only place `Deprecated` shows up is at the top of `query_test.exs` where it gets fed into `Postgrex.TypeModule.define/3`.

```
 ** (FunctionClauseError) no function clause matching in Deprecated.Postgrex.Extensions.Polygon/6
```

File: lib/postgrex/extensions/point.ex [hunk 3]
```diff
@@ -1,16 +1,20 @@
...
-
-  def encode(type_info, value, _, _) do
-    raise ArgumentError,
-      Postgrex.Utils.encode_msg(type_info, value, Postgrex.Point)
+  def encode(_) do
+    quote location: :keep do
+      %Postgrex.Point{x: x, y: y} ->
+        <<16::int32, x::float64, y::float64>>
+      other ->
+        Postgrex.Utils.encode_msg(other, Postgrex.Point)
```
Comment (dantswain) [inline lib/postgrex/extensions/point.ex]: Should be `raise ArgumentError, ...`

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**